src/Products/SiteAccess/VirtualHostMonster.py

```diff
@@ -148,7 +148,12 @@ def __call__(self, client, request, response=None):
                 protocol = stack.pop()
                 host = stack.pop()
                 if ':' in host:
-                    host, port = host.split(':')
+                    if host.startswith('['):
+                        # IPv6 address passed
+                        host, port = host.rsplit(':', 1)
```

Comment:
You're right about the IPv6-no-port scenario. It's too ugly to keep code that does not split correctly, even if the current implementation of `setServerURL` rectifies things.

However, I find overkill to use a regex.

I suugest:
```py
                host = stack.pop()
                if host.endswith(']'):
                    # IPv6 without port
                    request.setServerURL(protocol, host)
                else:
                    request.setServerURL(protocol, *host.rsplit(':', 1))
```

Reply:
Indeed the suggested regex is a little more complex. The intention is to always parse all parts in a single `match()` call, instead of needing one or two extra conditionals.

I don't mind reusing the deprecated `splitport()` if imported via `ZPublisher.HTTPRequest` as sugested by @jerome-nexedi.

Reply:
I'm also OK with @jmuchemb's suggestion, BTW.

Reply:
With `splitport` the implementation simplifies a lot, see https://github.com/zopefoundation/Zope/pull/314/commits/e6b0e76b1d276b108c82ba9cfbaadd86774801f5.